src/xmagics.hpp

```diff
@@ -0,0 +1,251 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Loic Gouarin and Sylvain Corlay       *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+#ifndef XMAGICS_HPP
+#define XMAGICS_HPP
+
+#include <algorithm>
+#include <cstdlib>
+#include <fstream>
+#include <memory>
+#include <regex>
+#include <type_traits>
+
+#include "cling/Interpreter/Value.h"
+#include "cling/Interpreter/Exception.h"
+#include "cling/Utils/Output.h"
+
+#include "xparser.hpp"
+
+namespace xeus
+{
+    struct XEUS_API xmagic_line
+    {
+        using type = xmagic_line;
+        virtual void operator()(const std::string& line) const{};
+    };
+    
+    struct XEUS_API xmagic_cell
+    {
+        virtual void operator()(const std::string& line, const std::string& cell) const{};
+    };
+
+    struct XEUS_API xmagic_line_cell: public xmagic_line, xmagic_cell
+    {
+    };
+
+    class XEUS_API xmagics_manager 
+    {
+    public:
+
+        xmagics_manager(){};
+
+        template<typename xmagic_type>
+        void register_magic(const std::string& magic_name, std::shared_ptr<xmagic_type> magic)
+        {
+            if (std::is_base_of<xmagic_line, xmagic_type>::value)
+                m_magic_line[magic_name] = std::dynamic_pointer_cast<xmagic_line>(magic);
+            if (std::is_base_of<xmagic_cell, xmagic_type>::value)
+                m_magic_cell[magic_name] = std::dynamic_pointer_cast<xmagic_cell>(magic);
+        }
+
+        void unregister_magic(const std::string& magic_name)
+        {
+            m_magic_cell.erase(magic_name);
+            m_magic_line.erase(magic_name);
+        }
+
+        bool find(const std::string& magic_name)
+        {
+            auto it = m_magic_cell.find(magic_name);
+            if (it != m_magic_cell.end())
+                return true;
+            return false;
```

Comment:
`return m_magic_cell.find(magic_name) != m_magic_cell.end();` is a shorter implementation.
Also consider renaming the function since `find` generally returns an iterator.